packages/next/client/components/reducer.ts

```diff
@@ -441,101 +506,132 @@ export function reducer(
         if (!cache.data) {
           cache.data = fetchServerResponse(url, state.tree)
         }
+
+        // readRoot to suspend here (in the reducer) until the fetch resolves.
         const flightData = cache.data.readRoot()
 
         // Handle case when navigating to page in `pages` from `app`
         if (typeof flightData === 'string') {
           return {
             canonicalUrl: flightData,
+            // Enable mpaNavigation
             pushRef: { pendingPush: true, mpaNavigation: true },
-            focusRef: { focus: false },
+            // Don't apply scroll and focus management.
+            focusAndScrollRef: { apply: false },
             cache: state.cache,
             tree: state.tree,
           }
         }
 
+        // Remove cache.data as it has been resolved at this point.
         cache.data = null
 
+        // TODO-APP: Currently the Flight data can only have one item but in the future it can have multiple paths.
         const flightDataPath = flightData[0]
 
+        // The one before last item is the router state tree patch
         const [treePatch] = flightDataPath.slice(-2)
-        const treePath = flightDataPath.slice(0, -3)
+
+        // Path without the last segment, router state, and the subTreeData
+        const flightSegmentPath = flightDataPath.slice(0, -3)
+
+        // Create new tree based on the flightSegmentPath and router state patch
         const newTree = walkTreeWithFlightDataPath(
           // TODO-APP: remove ''
-          ['', ...treePath],
+          ['', ...flightSegmentPath],
           state.tree,
           treePatch
         )
 
         mutable.previousTree = state.tree
         mutable.patchedTree = newTree
 
+        // Copy subTreeData for the root node of the cache.
         cache.subTreeData = state.cache.subTreeData
+        // Create a copy of the existing cache with the subTreeData applied.
         fillCacheWithNewSubTreeData(cache, state.cache, flightDataPath)
 
         return {
+          // Set href.
           canonicalUrl: href,
+          // Set pendingPush.
           pushRef: { pendingPush, mpaNavigation: false },
-          focusRef: { focus: true },
+          // All navigation requires scroll and focus management to trigger.
+          focusAndScrollRef: { apply: true },
+          // Apply patched cache.
           cache: cache,
+          // Apply patched tree.
           tree: newTree,
         }
       }
 
+      // This case should never be hit as `cacheType` is required and both cases are implemented.
       return state
     }
     case ACTION_SERVER_PATCH: {
       const { flightData, previousTree, cache } = action
+      // When a fetch is slow to resolve it could be that you navigated away while the request was happening or before the reducer runs.
+      // In that case opt-out of applying the patch given that the data could be stale.
       if (JSON.stringify(previousTree) !== JSON.stringify(state.tree)) {
```

Comment:
Note that JSON serialization is not stable, depending on the object construction the result of `JSON.stringify` may output different strings (but the same data).

Reply:
The tree is an array with fixed indexes so stringifying it results in the same string between runs, unlike objects with specific keys.